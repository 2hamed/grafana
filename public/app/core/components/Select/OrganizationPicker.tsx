import React, { Component } from 'react';
import _ from 'lodash';
import { getBackendSrv } from '@grafana/runtime';
import { AsyncSelect } from '@grafana/ui';
import { Organization } from 'app/types';

export interface OrgOption {
  id: number;
  value: string;
  label: string;
}

export interface Props {
  onSelected: (org: OrgOption) => void;
  className?: string;
}

export interface State {
  isLoading: boolean;
}

export class OrganizationPicker extends Component<Props, State> {
  debouncedSearch: any;

  constructor(props: Props) {
    super(props);
    this.state = { isLoading: false };
    this.search = this.search.bind(this);

    this.debouncedSearch = _.debounce(this.search, 300, {
      leading: true,
      trailing: true,
    });
  }

  async search(query?: string) {
    this.setState({ isLoading: true });

    if (_.isNil(query)) {
      query = '';
    }

    return await getBackendSrv()
      .get('/api/orgs', { query: query })
      .then((result: Organization[]) => {
        const organizations = result.map((org: Organization) => {
          return {
            id: org.id,
            value: org.name,
            label: org.name,
          };
        });

        this.setState({ isLoading: false });
        return organizations;
      });
  }

  render() {
    const { onSelected, className } = this.props;
    const { isLoading } = this.state;
    return (
      <div className="user-picker">
        <AsyncSelect
          inputId="org-picker"
          isLoading={isLoading}
          defaultOptions={true}
          loadOptions={this.debouncedSearch}
          onChange={onSelected}
          className={className}
          placeholder="Select an organization"
          noOptionsMessage={() => 'No organizations found'}
        />
      </div>
    );
  }
}
