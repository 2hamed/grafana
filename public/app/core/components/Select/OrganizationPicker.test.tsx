import React from 'react';
// @ts-ignore
import renderer from 'react-test-renderer';
import { OrganizationPicker } from './OrganizationPicker';

jest.mock('app/core/services/backend_srv', () => ({
  getBackendSrv: () => {
    return {
      get: () => {
        return Promise.resolve([]);
      },
    };
  },
}));

describe('OrganizationPicker', () => {
  it('renders correctly', () => {
    const props = {
      onSelected: () => {},
    };
    const tree = renderer.create(<OrganizationPicker {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
