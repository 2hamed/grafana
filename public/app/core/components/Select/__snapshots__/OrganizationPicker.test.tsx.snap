// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<div
  className="user-picker"
>
  <AsyncSelect
    autoFocus={false}
    backspaceRemovesValue={true}
    className=""
    components={Object {}}
    defaultOptions={true}
    inputId="org-picker"
    isClearable={false}
    isDisabled={false}
    isLoading={false}
    isMulti={false}
    isSearchable={true}
    loadOptions={[Function]}
    loadingMessage={[Function]}
    maxMenuHeight={300}
    noOptionsMessage={[Function]}
    onChange={[Function]}
    openMenuOnFocus={false}
    placeholder="Select an organization"
  />
</div>
`;
